iceoryx_examples/waitset/README.md

```diff
@@ -1,6 +1,586 @@
+ 1. The origin of the trigger, e.g. `this`
+ 2. A method of which can be called by the trigger to ask if it was triggered.
+ 3. A method which resets the trigger. Used when the WaitSet goes out of scope. 
+ 4. The id of the trigger.
+ 5. A callback with the signature `void (MyTriggerClass * )`.
+ 
+```cpp
+    iox::cxx::expected<iox::popo::WaitSetError>
+    attachTo(iox::popo::WaitSet& waitset,
+                    const MyTriggerClassEvents event,
+                    const uint64_t triggerId,
+                    const iox::popo::Trigger::Callback<MyTriggerClass> callback) noexcept
+    {
+        switch (event)
+        {
+        case MyTriggerClassEvents::PERFORMED_ACTION:
+        {
+            return waitset
+                .acquireTrigger(this,
+                                {this, &MyTriggerClass::hasPerformedAction},
+                                {this, &MyTriggerClass::invalidateTrigger},
+                                triggerId,
+                                callback)
+                .and_then([this](iox::popo::TriggerHandle& trigger) { 
+                    m_actionTrigger = std::move(trigger); });
+        }
+```
+
+When the parameter event has the value `ACTIVATE` we use the `isActivated` method
+for the trigger.
+```cpp
+        case MyTriggerClassEvents::ACTIVATE:
+        {
+            return waitset
+                .acquireTrigger(this,
+                                {this, &MyTriggerClass::isActivated},
+                                {this, &MyTriggerClass::invalidateTrigger},
+                                triggerId,
+                                callback)
+                .and_then([this](iox::popo::TriggerHandle& trigger) { 
+                    m_activateTrigger = std::move(trigger); });
+        }
+```
+
+The next thing on our checklist is the `invalidateTrigger` method used by the WaitSet
+to reset the _Trigger_ when it goes out of scope. Therefore we look up the
+correct trigger first by calling `isLogicalEqualTo` and then `reset` it.
+```cpp
+    void invalidateTrigger(const uint64_t uniqueTriggerId)
+    {
+        if (m_actionTrigger.getUniqueId() == uniqueTriggerId)
+        {
+            m_actionTrigger.invalidate();
+        }
+        else if (m_activateTrigger.getUniqueId() == uniqueTriggerId)
+        {
+            m_activateTrigger.invalidate();
+        }
+    }
+```
+
+#### Using MyTriggerClass
+
+The next thing we define is a free function, our `eventLoop`, which will handle
+all events of our waitset. The action is for every trigger the same, resetting
+the `MyTriggerClass` instance and call the callback which is attached to the
+trigger.
+```cpp
+void eventLoop()
+{
+    while (true)
+    {
+        auto triggerStateVector = waitset->wait();
+        for (auto& triggerState : triggerStateVector)
+        {
+            if (triggerState.getTriggerId() == ACTIVATE_ID)
+            {
+                triggerState.getOrigin<MyTriggerClass>()->reset(MyTriggerClassEvents::ACTIVATE);
```

Comment:
Why is the MyTriggerClass template argument needed? For an underlying cast to a pointer to this type?
Also, can we somehow get by without manual reset? I.e. if I wait again, I am only notified if the condition is true at or after the time of waiting, but not if it was before and switched back to false.

That would be closer to how condition variables behave.

Reply:
I stated the reason directly in the beginning of this Readme. The WaitSet is state based and as long no one changes the state of the WaitSet it is triggered again and again. Therefore we have to reset the event.

We want that the WaitSet behaves exactly in this manner so that it corresponds to the ros/dds version. 

But I am on your side, I would also prefer if the WaitSet would be only triggered once when the event has arrived and not all the time till someone handled the event and resets the state. @budrus any chance to convince you to agree to a change in behavior here?